src/amber/router/pipe/csrf.cr

```diff
@@ -22,20 +25,89 @@ module Amber
         !CHECK_METHODS.includes?(context.request.method)
       end
 
-      def valid_token?(context)
-        (context.params[PARAM_KEY]? || context.request.headers[HEADER_KEY]?) == self.class.token(context)
-      end
-
       def self.token(context)
-        context.session[CSRF_KEY] ||= Random::Secure.urlsafe_base64(32)
+        token_strategy.token(context)
       end
 
       def self.tag(context)
         %Q(<input type="hidden" name="#{PARAM_KEY}" value="#{token(context)}" />)
       end
 
       def self.metatag(context)
-        %Q(<meta name="#{HEADER_KEY}" content="#{token(context)}" />)
+        %Q(<meta name="#{PARAM_KEY}" content="#{token(context)}" />)
+      end
+
+      module BaseToken
+        def request_token(context)
+          context.params[PARAM_KEY]? || context.request.headers[HEADER_KEY]?
+        end
+
+        def real_session_token(context) : String
+          (context.session[CSRF_KEY] ||= Random::Secure.urlsafe_base64(TOKEN_LENGTH)).to_s
+        end
+      end
+
+      module RefreshableToken
+        extend self
+        extend BaseToken
+
+        def token(context) : String
+          real_session_token(context)
+        end
+
+        def valid_token?(context)
+          request_token = request_token(context)
+
+          request_token &&
+            Crypto::Subtle.constant_time_compare(request_token, token(context)) &&
+            !!context.session.delete(CSRF_KEY)
+        end
+      end
+
+      module PersistantToken
+        extend self
+        extend BaseToken
+
+        def valid_token?(context)
+          request_token = request_token(context)
+          return false if request_token.nil?
+          decoded_request_token = Base64.decode_string(request_token)
+          return false if decoded_request_token.bytesize != 2 * TOKEN_LENGTH
+          unmaseked_request_token = TokenOperations.unmask(decoded_request_token)
+          decoded_session_token = Base64.decode_string(real_session_token(context))
+
+          Crypto::Subtle.constant_time_compare(unmaseked_request_token, decoded_session_token)
+        end
+
+        def token(context) : String
+          unmask_token = Base64.decode_string(real_session_token(context))
+          masked_token = TokenOperations.mask(unmask_token)
+          Base64.urlsafe_encode(masked_token)
+        end
+
+        module TokenOperations
+          extend self
+
+          # Creates a masked version of the authenticity token that varies
+          # on each request. The masking is used to mitigate SSL attacks
+          # like BREACH.
+          def mask(unmasked_token)
+            one_time_pad = Bytes.new(TOKEN_LENGTH).tap { |buf| Random::Secure.random_bytes(buf) }
+            encrypted_csrf_token = xor_byte_strings(one_time_pad, unmasked_token.bytes)
+            "#{String.new(one_time_pad)}#{encrypted_csrf_token}"
```

Comment:
String.new is problematic as the bytes contained in `one_time_pad` aren't all ascii chars. We'll need to base64 encode it first.